samd21/clock: add xosc32/DFLL option

photonthunder opened the pull request:
Adds a clock option that uses the external 32kHz oscillator and DFLL to set GLCK0 to 48 MHz. [Issue #6261 ](https://github.com/RIOT-OS/RIOT/issues/6261)  is referenced for completeness.

thomaseichinger commented on boards/samd21-xpro/include/periph_conf.h:
```diff
@@ -70,6 +69,14 @@ extern "C" {
+#define CLOCK_DIV           (1U)
+#define CLOCK_CORECLOCK     (48000000U / CLOCK_DIV)
+#define CLOCK_XOSC32K       (32768UL)
+#define XTIMER_HZ			(CLOCK_XOSC32K)
```
Please align.

thomaseichinger commented on boards/samd21-xpro/include/periph_conf.h:
```diff
@@ -58,8 +59,6 @@ extern "C" {
  * CORECLOCK = 8MHz / DIV
  *
- * NOTE: A core clock frequency below 1MHz is not recommended
- *
```
Is this not true anymore?

thomaseichinger commented on boards/samr21-xpro/include/periph_conf.h:
```diff
@@ -58,26 +59,32 @@ extern "C" {
  * CORECLOCK = 8MHz / DIV
  *
- * NOTE: A core clock frequency below 1MHz is not recommended
- *
```
Not true anymore?

thomaseichinger commented on boards/samr21-xpro/include/periph_conf.h:
```diff
@@ -58,26 +59,32 @@ extern "C" {
  */
 #define CLOCK_USE_PLL       (1)
-
+    
```
Please remove trailing whitespaces.

thomaseichinger commented on boards/samr21-xpro/include/periph_conf.h:
```diff
@@ -58,26 +59,32 @@ extern "C" {
+#define CLOCK_DIV           (1U)
+#define CLOCK_CORECLOCK     (48000000U / CLOCK_DIV)
+#define CLOCK_XOSC32K       (32768UL)
+#define XTIMER_HZ			(CLOCK_XOSC32K)
```
Please align.

thomaseichinger commented on boards/samr21-xpro/include/periph_conf.h:
```diff
@@ -58,26 +59,32 @@ extern "C" {
 #endif
 /** @} */
-
+    
```
Whitespaces

thomaseichinger commented on cpu/samd21/cpu.c:
```diff
@@ -90,6 +105,75 @@ static void clk_init(void)
+    }
+
+    /* get the coarse and fine values stored in NVM (Section 9.3) */
+    uint32_t coarse = (*(uint32_t *)(0x806024) >> 26);  // Bits 63:58
```
Please use `/* C style comments */`

thomaseichinger commented on cpu/samd21/cpu.c:
```diff
@@ -90,6 +105,75 @@ static void clk_init(void)
+        /* Wait for DFLLLXXX sync */
+    }
+
+    //	/* select the DFLL as source for clock generator 0 (CPU core clock) */
```
Remove `//`.

thomaseichinger commented on cpu/samd21/cpu.c:
```diff
@@ -22,6 +22,19 @@
+#if CLOCK_USE_PLL
+#define CLOCK_8MHZ          1
+#define GEN2_ULP32K         1
+#endif
```
Couldn't this block be removed? `CLOCK_8MHZ` and `GEN2_ULP32K` get defined anyways in the following.

photonthunder commented:
@thomaseichinger thanks for the review.  I took out the "CORECLOCK below 1 MHz is not recommended" as a method of noting that in timer.c with the XOSC at 32 kHz the settings are below 1 MHz.  I believe the reason that you don't want to go below 1 MHz is so that xtimer_usleep will function?  If that is the case then I could turn on the 8 MHz clock and use that in timer.c.  In my case I just use the SysTick for any short delays since xtimer_usleep most likely won't yield anyway (if I understand things correctly).

photonthunder commented:
I re-thought the changes to timer.c and took them out.  Using the 8 MHz clock is good for the base samd21-xpro.

photonthunder commented:
squashed and rebased.  I want to pull all the clock generators into clock.c from rtc/rtt and so forth to better organize but I will save that for another PR in the future.

photonthunder commented on boards/samd21-xpro/include/periph_conf.h:
```diff
@@ -70,6 +71,12 @@ extern "C" {
 #define CLOCK_PLL_DIV       (1U)                /* adjust to your needs */
 /* generate the actual used core clock frequency */
 #define CLOCK_CORECLOCK     (((CLOCK_PLL_MUL + 1) * 1000000U) / CLOCK_PLL_DIV)
+#elif CLOCK_XOSC32_DFLL
```
updated

photonthunder commented:
I created issue #7208 to start a dialogue about how to clean up the generic clock generators. @thomaseichinger can I get you to review again?

photonthunder commented:
@thomaseichinger are you wanting wait on this one for the larger refractor or can we push this one through?

aabadie commented:
Tested this one on arduino-zero (also based on samd21) and it works after applying the required changes in the board config (just copy pasted from samd21-xpro and enabled CLOCK_USE_XOSC32_DFLL.

ACK on my side. @thomaseichinger  ?

photonthunder commented:
squashed and rebased

thomaseichinger commented:
@photonthunder I'm terribly sorry for the delay. Changes look good to me. Please rebase though.

photonthunder commented:
@thomaseichinger no worries... Rebased...

aabadie commented:
and go
